### ISSUE TITLE
Excessive memory consumption on Exq.Stats

### ISSUE THREAD

**Issue by pacoguzman** [akira/exq#350]
Sorry if this is not the place for this but I'm facing an excessive memory consumption on `Exq.Stats` process

```erlang
iex(my_app@10.73.215.121)3> :rpc.call(:"my_app@10.187.201.14", :sys, :get_state, [Exq.Stats])
%Exq.Stats.Server.State{queue: {[], []}, redis: Exq.Redis.Client}
iex(my_app@10.73.215.121)4> :rpc.call(:"my_app@10.187.201.14", :recon, :info, [Exq.Stats])   
[
  meta: [
    registered_name: Exq.Stats,
    dictionary: [
      "$initial_call": {Exq.Stats.Server, :init, 1},
      "$ancestors": [Exq.Sup,MyApp.Supervisor, #PID<35122.447.0>]
    ],
    group_leader: #PID<35122.446.0>,
    status: :waiting
  ],
  signals: [
    links: [#PID<35122.513.0>],
    monitors: [],
    monitored_by: [],
    trap_exit: true
  ],
  location: [
    initial_call: {:proc_lib, :init_p, 5},
    current_stacktrace: [
      {:gen_server, :loop, 7, [file: 'gen_server.erl', line: 381]},
      {:proc_lib, :init_p_do_apply, 3, [file: 'proc_lib.erl', line: 247]}
    ]
  ],
  memory_used: [
    memory: 1246410040,
    message_queue_len: 0,
    heap_size: 18481566,
    total_heap_size: 155801133,
    garbage_collection: [
      max_heap_size: %{error_logger: true, kill: true, size: 0},
      min_bin_vheap_size: 46422,
      min_heap_size: 233,
      fullsweep_after: 65535,
      minor_gcs: 1
    ]
  ],
  work: [reductions: 1408825464]
]
```

As you can see it takes around 1.2GB just after booting the application, we removed the Stats middleware but the problem is still there. Could you help us to find the issue?

**Comment by ananthakumaran** [comment]
how many jobs do you process per second? The stats server accumulates the events in a queue and sends the stats every second (stats_flush_interval). I would normally expect the message_queue to build up when the stats server can't keep up, but it looks empty `message_queue_len: 0`. Once the stats middleware is removed, none of this should matter, as nobody will send events to the stats server.

**Comment by pacoguzman** [comment]
Yes, I was able to follow the code to understand what it does, and you're correct. That's why our first option was to remove the middleware, as we measure processed jobs with other middleware through Prometheus. so I do not understand why this is still an issue.

**Comment by ananthakumaran** [comment]
```
[
  meta: [
    registered_name: PegasusExq.Stats,
    dictionary: [
      "$initial_call": {Exq.Stats.Server, :init, 1},
      "$ancestors": [PegasusExq.Sup, Pegasus.Supervisor, #PID<0.637.0>]
    ],
    group_leader: #PID<0.636.0>,
    status: :waiting
  ],
  signals: [
    links: [#PID<0.639.0>],
    monitors: [],
    monitored_by: [],
    trap_exit: true
  ],
  location: [
    initial_call: {:proc_lib, :init_p, 5},
    current_stacktrace: [
      {:gen_server, :loop, 7, [file: 'gen_server.erl', line: 403]},
      {:proc_lib, :init_p_do_apply, 3, [file: 'proc_lib.erl', line: 249]}
    ]
  ],
  memory_used: [
    memory: 8860,
    message_queue_len: 0,
    heap_size: 610,
    total_heap_size: 986,
    garbage_collection: [
      max_heap_size: %{error_logger: true, kill: true, size: 0},
      min_bin_vheap_size: 46422,
      min_heap_size: 233,
      fullsweep_after: 65535,
      minor_gcs: 21
    ]
  ],
  work: [reductions: 14490]
]
```

I just looked at my local stats server, one thing that seem to differ from your values is the `minor_gcs`. For me it increases as time goes by (it's already 26 now). So I wonder if it's anything to do with garbage collection? Do you run with any specific config? I would try to trigger garbage collection on that process and see how it goes.

**Comment by ananthakumaran** [comment]
I just checked another server which has been running for 3 days, the `minor_gcs` count is quite high
```
  memory_used: [
    memory: 18712,
    message_queue_len: 0,
    heap_size: 1598,
    total_heap_size: 2208,
    garbage_collection: [
      max_heap_size: %{error_logger: true, kill: true, size: 0},
      min_bin_vheap_size: 46422,
      min_heap_size: 233,
      fullsweep_after: 65535,
      minor_gcs: 15809
    ]
  ],
  work: [reductions: 689726861]
```

**Comment by pacoguzman** [comment]
We have multiple nodes, and on the other nodes the memory is low and the `minor_gcs` is higher than 1. But there is always one node with high memory and `minor_gcs` equals to one.

In the scenario, when we're not running the stats middleware, I guess the only code that is **call** is the `cleanup_host_stats` and `force_flush`. `force_flush` could be fast as the `queue` is empty we do not use the middleware to add data there, but I've found `cleanup_host_stats` is super slow. Could that function is a bottleneck and we call it a lot of times making the process consume more memory?

One notable thing is that each job contains args about 4K long and probably encoding/decoding could be expensive.

**Comment by ananthakumaran** [comment]
:) just came to the same conclusion.

BTW, do you have a constant set of hosts or you use something like kube where hostname is dynamic? Exq works well if the hostname of the workers is same upon restart. Otherwise, you will be leaving garbages and here I guess all the processes garbage is because the current worker only clean up its own upon restart, but if the hostname changes everytime, the list would keep growing and nobody would cleanup.

**Comment by pacoguzman** [comment]
That's it. We're using https://www.nomadproject.io as scheduler so you're right the hostnames change every now and then.

We've added a custom cleaner. Which runs on intervals just to check orphans queues and that process already run `cleanup_host_stats`. But I guess we started to run this too late.

Here is the code, I hope this doesn't scare you.

```elixir
defmodule Exq.Ephemeral.Cleaner do
  @moduledoc """
  When invoked this module cleanup state for ephemeral instances on an Exq
  running system.

  1. Re enqueue backups. So job backups for nodes that do not exist anymore
  will be routed to its original queue.
  2. Remove stats. We stored processed jobs per host that we need to remove for
  those hosts that do not exist anymore.

  Originally both tasks are executed on server bootup by node_identifier.
  """

  require Logger
  @name __MODULE__

  alias Exq.Redis.{Connection, JobQueue}

  def call(node_name) do
    Logger.info("#{inspect(@name)} starts Exq ephemeral cleanup")

    manager_name = Exq.Manager.Server.server_name(nil)
    exq_state = :sys.get_state(manager_name)
    node_ids = private_queue_identifiers(node_name)

    for {node_id, queue} <- backup_ids_and_queue(exq_state) do
      case Enum.member?(node_ids, node_id) do
        true ->
          Logger.info("#{inspect(@name)} exists node for backup_id: #{node_id} skipping")

        false ->
          Logger.info("#{inspect(@name)} does not exist node for backup_id: #{node_id} cleaning")

          # Cleanup Backups
          rescue_timeout(fn ->
            JobQueue.re_enqueue_backup(
              exq_state.redis,
              exq_state.namespace,
              node_id,
              queue
            )
          end)

          # Cleanup Processed
          rescue_timeout(fn ->
            Exq.Stats.Server.cleanup_host_stats(exq_state.stats, exq_state.namespace, node_id)
          end)
      end
    end
  end

  @prefix "queue:backup"
  defp backup_ids_and_queue(%{redis: redis, namespace: namespace}) do
    backup_key_pattern = JobQueue.full_key(namespace, "#{@prefix}::*")
    keys = Connection.keys!(redis, backup_key_pattern)

    Enum.map(keys, fn key ->
      [_, node_id, queue] = String.split(key, "::")
      {node_id, queue}
    end)
  end

  defp private_queue_identifiers(node_name) do
    Node.list()
    |> filter_by_node_name(node_name)
    |> map_by_node_identifier()
  end

  defp filter_by_node_name(nodes, node_name) do
    node_name_prefix = "#{node_name}@"

    Enum.filter(nodes, &String.starts_with?(Atom.to_string(&1), node_name_prefix))
  end

  defp map_by_node_identifier(nodes) do
    impl = Exq.Support.Config.node_identifier()
    func = :node_id
    args = []

    Enum.map(nodes, fn node ->
      if node == self() do
        apply(impl, func, args)
      else
        :rpc.call(node, impl, func, args)
      end
    end)
  end

  defp rescue_timeout(f) do
    rescue_timeout(nil, f)
  end

  defp rescue_timeout(fail_return, f) do
    try do
      f.()
    catch
      :exit, {:timeout, info} ->
        Logger.info("#{inspect(@name)} timeout occurred #{Kernel.inspect(info)}")
        fail_return
    end
  end
end
```

If you have any feedback it's more than welcome

**Comment by ananthakumaran** [comment]
You are well aware of the problem then. I don't have any experience with nomad. We use kube statefulset to handle this problem, which uses constant hostname.

**Comment by ananthakumaran** [comment]
fixed by #392
